cmd/oras/discover.go

```diff
@@ -0,0 +1,179 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras/cmd/oras/internal/option"
+
+	"github.com/need-being/go-tree"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+)
+
+type discoverOptions struct {
+	option.Common
+	option.Remote
+
+	targetRef    string
+	artifactType string
+	outputType   string
+}
+
+func discoverCmd() *cobra.Command {
+	var opts discoverOptions
+	cmd := &cobra.Command{
+		Use:   "discover [options] <name:tag|name@digest>",
+		Short: "discover artifacts from remote registry",
+		Long: `discover artifacts from remote registry
+
+Example - Discover all the artifacts linked with the specified reference:
+  oras discover localhost:5000/hello
+
+Example - Discover all the artifacts linked with the specified reference in a tree view:
+  oras discover localhost:5000/hello -o tree
+
+Example - Discover artifacts of type test-artifact test-artifact linked with the specified reference:
+  oras discover --artifact test-artifact localhost:5000/hello
+`,
+		Args: cobra.ExactArgs(1),
+		PreRun: func(cmd *cobra.Command, args []string) {
+			fmt.Println("Command discover is experimental and under development.")
```

Comment:
It's better to call it `preview` than experimental.

We also want to mark the `discover` command as `preview` in the `oras --help` command. Let's discuss offline for the temporary UX.

Reply:
Looks like there is no good way to mark a command as hidden in the `--help` outputs. I can think of two options 
- Add `preview` to the command short description
- Hide the command from oras `--help` output. (Chosen in current implementation)